### PR TITLE
Major trading engine and logic rework with precision, rounding and matching flow improvements

### PR DESCRIPTION
**Pull request by ysv**
- Remove rounding on order creation
- Validate volume and price precision on order creation
- Return invalid_price_or_volume error in case of fractional part overflow
- Validate Market min_price, max_price and min_amount depending on amount and price precision
- Market #min_amount_by_precision & #min_price_by_preciosion
- Market amount_precision + price_precision < FUNDS_PRECISION = 12
- Set Market FEE_PRECISION to 4
- Round estimated funds or order creation to FUNDS_PRECISION
- Matching::AbstractOrder interface
- Better Errors hierarchy in Matching module
- Validate Trade before publishing to TradeExecutor
- Rewrite Matching::Engine from scratch without recursion
- Remove legacy code which expects market orders to be in orderbook
- Return nil in MarketOrder #trade_with in case of not enough locked
- Raise MarketOrderbookError in case of market order in orderbook
- Update markets.yml.erb with new min_price, min_amount
- Migration for updating market min_price and min_amount
- Add Engine specs for market order with wrong estimates
- Update specs

### CODE REVIEW

**Comment by ysv** [comment]
TODO: 

- [x] Add funds, fee etc precision validations in TradeExecutor and somewhere else (task created)

- [x] OrderProcessor should process order after TE did it (task created)

- [x] min_price should be validated in the same way as min_amount

- [ ] sell market order doesn't work correctly

File: app/models/market.rb [hunk 1]
```diff
@@ -17,7 +17,14 @@
 
 class Market < ApplicationRecord
 
+  # Since we use decimal with 16 digits fractional part for storing numbers in DB
+  # sum of multipliers fractional parts must not be greater then 16.
+  # In the wors situation we have 3 multipliers (price * amount * fee).
```
Comment (calj) [inline app/models/market.rb]: worse instead of wors

Comment (ysv) [reply app/models/market.rb]: Thanks

File: app/trading/matching/engine.rb [hunk 2]
```diff
@@ -63,54 +39,136 @@ def enqueue(*args)
...
-      return if order.filled?
-      return unless (counter_order = counter_book.top)
+    def match(order)
+      book, opposite_book = orderbook.get_books(order.type)
+
+      loop do
+        # If order is fulfilled we stop matching.
+        break if order.filled?
+
+        # If oposite orderbook is empty:
```
Comment (mnaichuk) [inline app/trading/matching/engine.rb]: opposite

**Pull request merged**